src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/ComplexTypeModelBinder.cs

```diff
@@ -2,17 +2,36 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.Collections.Generic;
 using System.Reflection;
 using System.Threading.Tasks;
 using Microsoft.AspNetCore.Mvc.Internal;
 
-namespace Microsoft.AspNetCore.Mvc.ModelBinding
+namespace Microsoft.AspNetCore.Mvc.ModelBinding.Binders
 {
     /// <summary>
-    /// <see cref="IModelBinder"/> implementation for binding complex values.
+    /// <see cref="IModelBinder"/> implementation for binding complex types.
     /// </summary>
-    public class MutableObjectModelBinder : IModelBinder
+    public class ComplexTypeModelBinder : IModelBinder
     {
+        private readonly Dictionary<ModelMetadata, IModelBinder> _propertyBinders;
+
+        /// <summary>
+        /// Creates a new <see cref="ComplexTypeModelBinder"/>.
+        /// </summary>
+        /// <param name="propertyBinders">
+        /// The <see cref="IDictionary{TKey, TValue}"/> of binders to use for binding properties.
+        /// </param>
+        public ComplexTypeModelBinder(IDictionary<ModelMetadata, IModelBinder> propertyBinders)
+        {
+            if (propertyBinders == null)
+            {
+                throw new ArgumentNullException(nameof(propertyBinders));
+            }
+
+            _propertyBinders = new Dictionary<ModelMetadata, IModelBinder>(propertyBinders);
```

Comment:
The only reason is to get the field typed as `Dictionary` and avoid enumerator alloc. Open to suggestions.